doc: Bluetooth: Documentation overhaul

carlescufi opened the pull request:
Overhaul the Bluetooth documentation to split it into manageable units
and include additional information, such as architecture and tooling.

Fixes #13394

carlescufi committed:
doc: Bluetooth: Documentation overhaul

Overhaul the Bluetooth documentation to split it into manageable units
and include additional information, such as architecture and tooling.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

jhedberg commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,197 @@
+  choose to use whatever Host running on any platform they prefer. For example,
+  the host can be the Linux BLE Host stack (BlueZ) running on any processor
+  capable of supporting Linux. The Host processor may also run Zephyr and the
+  Zephyr OS BLE Host.
```
Might want to mention here the case of combining Zephyr host with non-Zephyr controller, e.g. the case with qemu or native posix using your PC's internal Bluetooth adapter.

carlescufi replied:
Added thanks

jhedberg reviewed (commented):
Really impressive start! I can see you spent a lot of time on this, thanks! Just one minor comment from my side.

aescolar commented on doc/guides/bluetooth/index.rst:
```diff
@@ -3,310 +3,18 @@
-
-The PICS features for each supported protocol & profile can be found in
-the following documents:
+Zephyr includes a complete Bluetooth Low Energy stack from application to radio
```
Nit: I wouldn't say "application" (even though there is a few sample apps.

carlescufi replied:
I would actually prefer to leave as-is. We do provide some apps that, combined, can make a real application.

aescolar commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,88 @@
+
+      * Features, buffer sizes/counts, stack sizes, etc.
+  
+  * Portable to all architectures supported by Zephyr (including big and
```
Silly comment: The big endian support is work in progress ;)

carlescufi replied:
we'll consider it almost done ;)

aescolar replied:
Until #13229 is merged we are a bit far from done :)

aescolar commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,88 @@
+
+* Bluetooth 5.0 compliant (ESR10)
+
+  * Bluetooth subsystem and infrastructure
```
This point seems out of place in this list

carlescufi replied:
removed

aescolar commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,88 @@
+  * Unlimited role and connection count, all roles supported
+  * Concurrent multi-protocol support ready
+  * Intelligent scheduling of roles to minimize overlap
+  * Portable design to any open BLE radio, currently supports Nordic
```
This part will only be true after 1.14, so maybe we should not advertise it yet, or?

carlescufi replied:
It's true in master, this doc is in master so I think it's fine

aescolar reviewed (commented):
Thanks Carles, very nice additions.
Just a few minor comments

Commit:
doc: Bluetooth: Add initial guide for the host stack

Add initial sections for the Bluetooth host stack, as well as a simple
diagram of the various protocol & profile layers.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Vudentz committed:
doc: Bluetooth: Add initial documentation for the L2CAP layer

Add initial sections for L2CAP layer.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
doc: Bluetooth: Add initial documentation for the GATT layer

Add initial sections for GATT layer.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Commit:
doc: Bluetooth: use :cpp:func: instead of :c:func:

:c:func: doesn't seem to generate proper hyperlinks, whereas
:cpp:func: works as expected.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

pskhansen reviewed (approved):
Just a few comments.

carlescufi commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,197 @@
+  the multiple (non real-time) network and transport protocols enabling
+  applications to communicate with peer devices in a standard and interoperable
+  way
+* **Controller**: The Controller implements either the Link Layer (LE LL), the
```
Fixed thanks

carlescufi commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,197 @@
+Host Controller Interface
+=========================
+
+The Bluetooth Specification describes the format in which a Host must
```
Yep, done thanks.

carlescufi commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,197 @@
+BLE-enabled builds that can be produced from the Zephyr project codebase:
+
+* **Controller-only build**: When built as a BLE Controller, Zephyr project
+  contains the Link Layer and a special application (
```
Done thanks

carlescufi commented on doc/guides/bluetooth/index.rst:
```diff
@@ -3,310 +3,18 @@
-    :local:
-    :depth: 2
+The Bluetooth section contains information regarding the Bluetooth stack of the
+Zephyr kernel. Use the information to understand the principles behind the
```
fixed

carlescufi commented on doc/guides/bluetooth/index.rst:
```diff
@@ -3,310 +3,18 @@
-.. contents::
-    :local:
-    :depth: 2
+The Bluetooth section contains information regarding the Bluetooth stack of the
```
fixed

carlescufi commented on doc/guides/bluetooth/bluetooth-dev.rst:
```diff
@@ -0,0 +1,63 @@
+Hardware setup
+**************
+
+This section describes the options you have when building and debuggin Bluetooth
```
Fixed, thanks

carlescufi commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,370 @@
+  Bluetooth specification does not specify how HCI is implemented in this
+  single-chip configuration and so how HCI commands, events and data flows between
+  the two can be implementation-specific. The Zephyr OS supports this
+  configuration on Nordic Semiconductor ICs, for those applications and designs
```
Yes, I agree. I will rework that

carlescufi replied:
Done

dbkinder commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,540 @@
+:cpp:func:`bt_gatt_discover` API which takes the
+:cpp:class:`bt_gatt_dicover_params` struct which describes the type of
+discovery. The parameters also serves as a filter when setting the ``uuid``
+field only attributes which matches will be discovered, by contranst setting it
```
```suggestion
field only attributes which matches will be discovered. In contrast, setting it
```

dbkinder commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,540 @@
+:cpp:class:`bt_gatt_dicover_params` struct which describes the type of
+discovery. The parameters also serves as a filter when setting the ``uuid``
+field only attributes which matches will be discovered, by contranst setting it
+to NULL makes all attributes to be discovered.
```
```suggestion
to NULL allows all attributes to be discovered.
```

dbkinder commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,540 @@
+
+Standard
+========
+
```
Work in progress, right?

carlescufi replied:
Yep, will be a follow-up PR

dbkinder commented on doc/guides/bluetooth/bluetooth-dev.rst:
```diff
@@ -0,0 +1,124 @@
+All the :ref:`bluetooth-configs` and :ref:`bluetooth-build-types` are supported
+but you might need to build Zephyr more than once if you are using a dual-chip
+configuration or if you have multiple cores in your SoC each running a different
+build type (eg. one running the Host, the other the Controller).
```
```suggestion
build type (e.g., one running the Host, the other the Controller).
```

dbkinder commented on doc/guides/bluetooth/bluetooth-tools.rst:
```diff
@@ -0,0 +1,191 @@
+
+   ExecStart=/usr/libexec/bluetooth/bluetoothd -E
+
+Finally you can reload and restart the daemon:
```
```suggestion
Finally, reload and restart the daemon:
```

dbkinder commented on doc/guides/bluetooth/bluetooth-tools.rst:
```diff
@@ -0,0 +1,191 @@
+* To a serial port in Native POSIX through the use of a command-line option
+  passed to the Native POSIX executable: ``--bt-dev=hci0``
+
+On the host side, BlueZ allows to export its Bluetooth controller
```
```suggestion
On the host side, BlueZ allows you to export its Bluetooth controller
```

dbkinder commented on doc/guides/bluetooth/img/ble_cfg_dual.xml:
```diff
@@ -0,0 +1 @@
+<mxfile modified="2019-03-07T14:48:10.055Z" host="www.draw.io" agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36" version="10.3.5" etag="e-D3ClfASJyunx5gBdFe" type="dropbox"><diagram id="3d368da2-d055-8685-6f25-468c73dcbdd0" name="Page-1">7Vtbc+I2FP41zGwfkvEdeAQ2l52mncymmXb3pWOwAE2E5foSSH/9HtkSWJYCprG5dGFnFuvIlq3v3L+Yjj1arO5 ...
```
Thanks for including these drawing source files!  We may need to teach checkpatch not to worry about "misspellings" in XML files @nashif

dbkinder commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,98 @@
+    :local:
+    :depth: 2
+
+Zephyr has had a strong focus on Bluetooth and, in particular, on Bluetooth
```
I think this would read better, and clarify whos inception you're talking about:
```
Since its inception, Zephyr has had a strong focus on Bluetooth and, in particular, on Bluetooth
Low Energy (BLE).
```

dbkinder commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,98 @@
+
+Zephyr has had a strong focus on Bluetooth and, in particular, on Bluetooth
+Low Energy (BLE) since its inception. Through the contributions of several
+companies and individuals both involved in existing open source
```
```suggestion
companies and individuals involved in existing open source
```

dbkinder commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,98 @@
+
+  * Support for all combinations of Host and Controller builds:
+
+    * Controller-only (HCI) over UART, SPI and USB physical transports
```
```suggestion
    * Controller-only (HCI) over UART, SPI, and USB physical transports
```

dbkinder commented on doc/guides/bluetooth/overview.rst:
```diff
@@ -0,0 +1,98 @@
+  * Support for all combinations of Host and Controller builds:
+
+    * Controller-only (HCI) over UART, SPI and USB physical transports
+    * Host-only over UART, SPI and IPM (shared memory)
```
```suggestion
    * Host-only over UART, SPI, and IPM (shared memory)
```

dbkinder commented on doc/scripts/extract_content.py:
```diff
@@ -149,6 +149,8 @@ def extract_content(content):
     # modification times unnecessarily, as this triggers documentation
     # rebuilds.
     for output in content.outputs:
+        if 'guides/bluetooth' in output.src:
```
What is this about?

carlescufi replied:
oops that is test code, I will remove, sorry.

dbkinder reviewed (dismissed):
Thanks for the doc updates!

carlescufi commented:
@dbkinder we've addressed the issues

Vudentz reviewed (approved):
LGTM

jhedberg commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -0,0 +1,539 @@
+
+.. note::
+  The ``recv`` callback is called directly from RX Thread thus it is not
+  allowed to block.
```
This isn't completely true. You can (and should) block for small periods of time if it simplifies the code. This is e.g. why it's possible to do `bt_hci_cmd_send_sync()` from within the RX thread.

jhedberg replied:
So I'd change this to something like:

"The recv callback is called from the RX thread, so long blocking operations should be avoided and all blocking operations which rely on the RX thread itself must not be done since it would result in a deadlock."

Vudentz replied:
I will update the text, but in case of bt_l2cap_chan.recv it is a callback on the application, well appart from fixed channels, so I guess it is useful to inform in what context they will get called.

jhedberg reviewed (approved):
Consider this a "conditionally approved", @Vudentz please make sure that my RX thread comment gets taken care of in follow-up PRs.
